File: app/src/main/java/me/imzack/app/ender/view/adapter/TypePickerForPlanMigrationGridAdapter.java
package me.imzack.app.ender.view.adapter;

import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import me.imzack.app.ender.R;
import me.imzack.app.ender.model.DataManager;
import me.imzack.app.ender.model.bean.Type;
import me.imzack.app.ender.util.ResourceUtil;
import me.imzack.app.ender.util.StringUtil;
import me.imzack.app.ender.view.widget.CircleColorView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class TypePickerForPlanMigrationGridAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Type> mExcludedTypeList;
    private OnItemClickListener mOnItemClickListener;

    public TypePickerForPlanMigrationGridAdapter(DataManager dataManager, String excludedTypeCode) {
        mExcludedTypeList = dataManager.getExcludedTypeList(excludedTypeCode);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ItemViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_grid_type_picker_for_plan_migration, parent, false));
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        ItemViewHolder itemViewHolder = (ItemViewHolder) holder;
        Type type = mExcludedTypeList.get(position);
        setTypeMarkIcon(itemViewHolder.mTypeMarkIcon, type.getTypeMarkColor(), type.getTypeMarkPattern(), type.getTypeName());
        setTypeNameText(itemViewHolder.mTypeNameText, type.getTypeName());
        setItemView(itemViewHolder);
    }

    @Override
    public int getItemCount() {
        return mExcludedTypeList.size();
    }

    private void setTypeMarkIcon(CircleColorView typeMarkIcon, String typeMarkColor, String typeMarkPattern, String typeName) {
        typeMarkIcon.setFillColor(Color.parseColor(typeMarkColor));
        typeMarkIcon.setInnerIcon(ResourceUtil.getDrawable(typeMarkPattern));
        typeMarkIcon.setInnerText(StringUtil.getFirstChar(typeName));
    }

    private void setTypeNameText(TextView typeNameText, String typeName) {
        typeNameText.setText(typeName);
    }

    private void setItemView(final ItemViewHolder itemViewHolder) {
        itemViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mOnItemClickListener != null) {
                    Type type = mExcludedTypeList.get(itemViewHolder.getLayoutPosition());
                    mOnItemClickListener.onItemClick(type.getTypeCode(), type.getTypeName());
                }
            }
        });
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.ic_type_mark)
        CircleColorView mTypeMarkIcon;
        @BindView(R.id.text_type_name)
        TextView mTypeNameText;

        public ItemViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    public interface OnItemClickListener {
        void onItemClick(String typeCode, String typeName);
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        mOnItemClickListener = listener;
    }
}

File: app/src/main/kotlin/me/imzack/app/end/view/adapter/TypePickerForPlanMigrationGridAdapter.kt
package me.imzack.app.end.view.adapter

import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_grid_type_picker_for_plan_migration.*
import me.imzack.app.end.R
import me.imzack.app.end.model.DataManager
import me.imzack.app.end.util.ResourceUtil
import me.imzack.app.end.util.StringUtil
import me.imzack.app.end.view.widget.CircleColorView

class TypePickerForPlanMigrationGridAdapter(excludedTypeCode: String) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    var mOnItemClickListener: ((typeCode: String, typeName: String) -> Unit)? = null

    private val mExcludedTypeList = DataManager.getExcludedTypeList(excludedTypeCode)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            ItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_grid_type_picker_for_plan_migration, parent, false))

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val itemViewHolder = holder as ItemViewHolder
        val (_, name, markColor, markPattern, _, hasMarkPattern) = mExcludedTypeList[position]
        setTypeMarkIcon(itemViewHolder.ic_type_mark, markColor, hasMarkPattern, markPattern, name)
        setTypeNameText(itemViewHolder.text_type_name, name)
        setItemView(itemViewHolder)
    }

    override fun getItemCount() = mExcludedTypeList.size

    private fun setTypeMarkIcon(typeMarkIcon: CircleColorView, typeMarkColor: String, hasTypeMarkPattern: Boolean, typeMarkPattern: String?, typeName: String) {
        typeMarkIcon.setFillColor(Color.parseColor(typeMarkColor))
        typeMarkIcon.setInnerIcon(if (hasTypeMarkPattern) ResourceUtil.getDrawable(typeMarkPattern!!) else null)
        typeMarkIcon.setInnerText(StringUtil.getFirstChar(typeName))
    }

    private fun setTypeNameText(typeNameText: TextView, typeName: String) {
        typeNameText.text = typeName
    }

    private fun setItemView(itemViewHolder: ItemViewHolder) {
        itemViewHolder.itemView.setOnClickListener {
            val (code, name) = mExcludedTypeList[itemViewHolder.layoutPosition]
            mOnItemClickListener?.invoke(code, name)
        }
    }

    class ItemViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView), LayoutContainer
}
